### PR TITLE
feat(wash)!: Adds support for fetching dependencies automatically

### PR DESCRIPTION
**Pull request by thomastaylor312**
This uses the newly available wit commands from wasm-pkg-tools to fetch things from upstream registry. I have only updated select examples (that are less likely to be used while we finish the last bits of work) so we could test things.

I also fixed the wash build tests so they could run properly in parallel (which hopefully makes them faster too). Once we have all wasi cloud wit in OCI, we can update all deps and roll to a new version of wash. This will involve updating git ignores to ignore the deps directory, but leaving the current deps intact for those on older versions of wash for at least the next version or two. Once we're past that we can remove the deps dirs

### CODE REVIEW

**Comment by thomastaylor312** [comment]
This should be ready for review. Please note that until we push wasi:logging (probably tomorrow morning) the tests will fail here, and I'll need to remove the patch section as soon as I cut wkg 0.7.4 (once again, probably tomorrow morning).

File: .github/workflows/examples-components.yml
Comment (brooksmtownsend) [inline .github/workflows/examples-components.yml:49]: Don't forget or you will be forsworn!

File: Cargo.toml
Comment (brooksmtownsend) [inline Cargo.toml:393]: I think I saw this reflected in the PR, but just make sure we get these removed before merge

File: crates/wash-cli/src/bin/wash.rs
Comment (brooksmtownsend) [inline crates/wash-cli/src/bin/wash.rs:245]: We should audit our top level commands soon ™️

File: crates/wash-cli/src/wit.rs
Comment (brooksmtownsend) [inline crates/wash-cli/src/wit.rs:48]: nit: Other places in wash we've called this a destination / output. I think this is fine, but we should make them consistent at least in a followup. (this is a good first issue)

File: crates/wash-cli/src/wit.rs
Comment (brooksmtownsend) [inline crates/wash-cli/src/wit.rs:90]: Do we need to `flush` the lock_file here to doubley be sure it writes?

File: crates/wash-cli/src/wit.rs
Comment (brooksmtownsend) [inline crates/wash-cli/src/wit.rs:102]: Same comment re: flush

File: Cargo.toml
Comment (brooksmtownsend) [inline Cargo.toml:228]: can we kill the `homedir` crate if we have this?

**Review by brooksmtownsend — COMMENTED** [review]
All nits/comments/questions, but largely looks great to me. Leaving it unapproved for now to avoid accidental merge

File: Cargo.toml
Comment (thomastaylor312) [reply Cargo.toml:228]: Oh I was going to and forgot. Thanks for the reminder

File: crates/wash-cli/src/wit.rs
Comment (thomastaylor312) [reply crates/wash-cli/src/wit.rs:48]: I was trying to mostly match `wkg wit` here. But I did have to rename the arg anyway because we already have an "output"

**Review by brooksmtownsend — COMMENTED** [review]
LGTM! Requesting we don't merge this until we cut wash 0.35 just to keep the list of changes somewhat minor

**Review by brooksmtownsend — APPROVED** [review]
LGTM! Requesting we don't merge this until we cut wash 0.35 just to keep the list of changes somewhat minor

**Pull request merged**
